File: mem0/vector_stores/upstash_vector.py
import logging
from typing import Dict, List, Optional

from pydantic import BaseModel
from upstash_vector import Index

from mem0.vector_stores.base import VectorStoreBase

logger = logging.getLogger(__name__)


class OutputData(BaseModel):
    id: Optional[str]  # memory id
    score: Optional[float]  # is None for `get` method
    payload: Optional[Dict]  # metadata


class UpstashVector(VectorStoreBase):
    def __init__(
        self,
        url: Optional[str] = None,
        token: Optional[int] = None,
        client: Optional[Index] = None,
        namespace: Optional[str] = None,
    ):
        """
        Initialize the UpstashVector vector store.

        Args:
            url (str, optional): URL for Upstash Vector index. Defaults to None.
            token (int, optional): Token for Upstash Vector index. Defaults to None.
            client (Index, optional): Existing `upstash_vector.Index` client instance. Defaults to None.
            namespace (str, optional): Default namespace for the index. Defaults to None.
        """
        if client:
            self.client = client
        elif url and token:
            self.client = Index(url, token)
        else:
            raise ValueError("Either a client or URL and token must be provided.")

        if namespace is None:
            self.namespace = ""
        else:
            self.namespace = namespace

    def create_col(self):
        """
        Upstash Vector does not have collections. A single vector database only contains one size of vectors.
        """

        raise NotImplementedError(
            "Upstash Vector does not have collections. A single vector database only contains one size of vectors."
        )

    def insert(self, vectors: list, payloads: list = None, ids: list = None):
        """
        Insert vectors

        Args:
            vectors (list): List of vectors to insert.
            payloads (list, optional): List of payloads corresponding to vectors. These will be passed as metadatas to the Upstash Vector client. Defaults to None.
            ids (list, optional): List of IDs corresponding to vectors. Defaults to None.
        """
        logger.info(f"Inserting {len(vectors)} vectors into namespace {self.namespace}")
        processed_vectors = [
            {
                "id": ids[i] if ids else None,
                "vector": vectors[i],
                "metadata": payloads[i] if payloads else None,
            }
            for i, v in enumerate(vectors)
        ]
        self.client.upsert(
            vectors=processed_vectors,
            namespace=self.namespace,
        )

    def search(self, query: list, limit: int = 5, filters: Dict = None) -> List[OutputData]:
        """
        Search for similar vectors.

        Args:
            query (list): Query vector.
            limit (int, optional): Number of results to return. Defaults to 5.
            filters (Dict, optional): Filters to apply to the search.

        Returns:
            List[OutputData]: Search results.
        """

        def stringify(x):
            return f'"{x}"' if isinstance(x, str) else x

        filters_str = " AND ".join([f"{k} = {stringify(v)}" for k, v in filters.items()]) if filters else None

        response = self.client.query(
            vector=query,
            top_k=limit,
            filter=filters_str,
            include_metadata=True,
            namespace=self.namespace,
        )
        print("res", response)
        return [
            OutputData(
                id=res.id,
                score=res.score,
                payload=res.metadata,
            )
            for res in response
        ]

    def delete(self, vector_id: int):
        """
        Delete a vector by ID.

        Args:
            vector_id (int): ID of the vector to delete.
        """
        self.client.delete(
            ids=[vector_id],
            namespace=self.namespace,
        )

    def update(self, vector_id: int, vector: list = None, payload: dict = None):
        """
        Update a vector and its payload.

        Args:
            vector_id (int): ID of the vector to update.
            vector (list, optional): Updated vector. Defaults to None.
            payload (dict, optional): Updated payload. Defaults to None.
        """
        self.client.update(
            id=vector_id,
            vector=vector,
            metadata=payload,
            namespace=self.namespace,
        )

    def get(self, vector_id: int) -> dict:
        """
        Retrieve a vector by ID.

        Args:
            vector_id (int): ID of the vector to retrieve.

        Returns:
            dict: Retrieved vector.
        """
        response = self.client.fetch(
            ids=[vector_id],
            namespace=self.namespace,
            include_metadata=True,
        )
        print(response, response[0].id)
        if len(response) == 0:
            return None
        return OutputData(
            id=response[0].id,
            score=None,
            payload=response[0].metadata,
        )

    def list_cols(self) -> list:
        """
        Upstash Vector does not have collections. A single vector database only contains one size of vectors.
        """
        raise NotImplementedError(
            "Upstash Vector does not have collections. A single vector database only contains one size of vectors."
        )

    def delete_col(self):
        """
        Upstash Vector does not have collections. A single vector database only contains one size of vectors.
        """
        raise NotImplementedError(
            "Upstash Vector does not have collections. A single vector database only contains one size of vectors."
        )

    def col_info(self) -> dict:
        """
        Upstash Vector does not have collections. A single vector database only contains one size of vectors.
        """
        raise NotImplementedError(
            "Upstash Vector does not have collections. A single vector database only contains one size of vectors."
        )

File: mem0-ts/src/oss/src/vector_stores/upstash_vector.ts
import type { Index, QueryResult, Vector } from "@upstash/vector";
import { VectorStore } from "./base";
import { SearchFilters, VectorStoreConfig, VectorStoreResult } from "../types";
import { loadPeer } from "../utils/load_peer";

interface UpstashVectorConfig extends VectorStoreConfig {
  collectionName: string;
  url?: string;
  token?: string;
  /** Pre-configured Upstash Vector client instance (typed as `any` to keep
   *  the optional driver's types out of the published type declarations). */
  client?: any;
}

type UpstashMetadata = Record<string, unknown>;

export class UpstashVector implements VectorStore {
  private client!: Index<UpstashMetadata>;
  private readonly config: UpstashVectorConfig;
  private readonly collectionName: string;

  constructor(config: UpstashVectorConfig) {
    if (!config.collectionName) {
      throw new Error("collectionName is required for Upstash Vector.");
    }
    if (!config.client && !(config.url && config.token)) {
      throw new Error("Either a client or url and token must be provided.");
    }

    this.config = config;
    this.collectionName = config.collectionName;
  }

  /**
   * Lazily construct (or reuse) the Upstash Vector client, importing the
   * optional `@upstash/vector` peer only when the store is first used so
   * consumers that never touch Upstash Vector don't need it installed.
   */
  private async ensureClient(): Promise<void> {
    if (this.client) return;

    const config = this.config;
    if (config.client) {
      this.client = config.client;
    } else {
      const sdk = await loadPeer(
        "@upstash/vector",
        "Upstash Vector store",
        () => import("@upstash/vector"),
      );
      this.client = new sdk.Index({
        url: config.url,
        token: config.token,
      });
    }
  }

  async initialize(): Promise<void> {
    await this.ensureClient();
  }

  async insert(
    vectors: number[][],
    ids: string[],
    payloads: Record<string, any>[],
  ): Promise<void> {
    await this.initialize();
    const upsertData = vectors.map((vector, idx) => {
      return {
        id: ids[idx],
        vector,
        metadata: payloads[idx] ?? {},
      };
    });

    await this.client.upsert(upsertData, { namespace: this.collectionName });
  }

  async search(
    query: number[],
    topK: number = 5,
    filters?: SearchFilters,
  ): Promise<VectorStoreResult[]> {
    await this.initialize();
    const response = await this.client.query<UpstashMetadata>(
      {
        vector: query,
        topK,
        filter: this.convertFilters(filters),
        includeMetadata: true,
      },
      { namespace: this.collectionName },
    );

    return response.map((result) => this.parseResult(result));
  }

  async keywordSearch(
    query: string,
    topK: number = 5,
    filters?: SearchFilters,
  ): Promise<VectorStoreResult[] | null> {
    await this.initialize();
    try {
      const response = await this.client.query<UpstashMetadata>(
        {
          data: query,
          topK,
          filter: this.convertFilters(filters),
          includeMetadata: true,
        },
        { namespace: this.collectionName },
      );

      return response.map((result) => this.parseResult(result));
    } catch (error) {
      console.error(`Error during keyword search for query '${query}':`, error);
      return null;
    }
  }

  async get(vectorId: string): Promise<VectorStoreResult | null> {
    await this.initialize();
    const response = await this.client.fetch<UpstashMetadata>([vectorId], {
      includeMetadata: true,
      namespace: this.collectionName,
    });
    const vector = response[0];

    if (!vector) {
      return null;
    }

    return {
      id: String(vector.id),
      payload: (vector.metadata ?? {}) as Record<string, any>,
    };
  }

  async update(
    vectorId: string,
    vector: number[],
    payload: Record<string, any>,
  ): Promise<void> {
    await this.initialize();
    // Upstash's `update` can't set the vector and metadata in one call (its
    // payload is a discriminated union of vector | data | metadata), so a
    // single `upsert` replaces both atomically, the same way insert() writes.
    await this.client.upsert(
      {
        id: vectorId,
        vector,
        metadata: payload,
      },
      { namespace: this.collectionName },
    );
  }

  async delete(vectorId: string): Promise<void> {
    await this.initialize();
    await this.client.delete(vectorId, { namespace: this.collectionName });
  }

  async deleteCol(): Promise<void> {
    await this.initialize();
    await this.client.reset({ namespace: this.collectionName });
  }

  async list(
    filters?: SearchFilters,
    topK: number = 100,
  ): Promise<[VectorStoreResult[], number]> {
    await this.initialize();
    const results: VectorStoreResult[] = [];
    let cursor = "0";

    do {
      const response = await this.client.range<UpstashMetadata>(
        {
          cursor,
          limit: Math.min(100, topK - results.length),
          includeMetadata: true,
        },
        { namespace: this.collectionName },
      );

      for (const vector of response.vectors) {
        if (this.matchesFilters(vector, filters)) {
          results.push({
            id: String(vector.id),
            payload: (vector.metadata ?? {}) as Record<string, any>,
          });
        }

        if (results.length >= topK) {
          break;
        }
      }

      cursor = response.nextCursor;
      // Upstash returns an empty-string cursor once the scan is exhausted (it
      // never comes back as "0"), so "" is the termination sentinel. Checking
      // for "0" here would re-scan from the start and return duplicates.
    } while (cursor !== "" && results.length < topK);

    return [results, results.length];
  }

  async getUserId(): Promise<string> {
    return "anonymous-upstash-vector";
  }

  async setUserId(): Promise<void> {
    return;
  }

  async reset(): Promise<void> {
    await this.deleteCol();
  }

  private parseResult(result: QueryResult<UpstashMetadata>): VectorStoreResult {
    return {
      id: String(result.id),
      payload: (result.metadata ?? {}) as Record<string, any>,
      score: result.score,
    };
  }

  private stringifyFilterValue(value: unknown): string {
    if (typeof value === "string") {
      return JSON.stringify(value);
    }

    if (typeof value === "boolean") {
      return value ? "true" : "false";
    }

    return String(value);
  }

  private convertFilters(filters?: SearchFilters): string | undefined {
    if (!filters) {
      return undefined;
    }

    const clauses: string[] = [];
    for (const [key, value] of Object.entries(filters)) {
      if (value === undefined || value === null) continue;
      const clause = this.buildFilterClause(key, value);
      if (clause) clauses.push(clause);
    }
    return clauses.length > 0 ? clauses.join(" AND ") : undefined;
  }

  // mem0's universal comparison operators -> Upstash filter tokens.
  private static readonly COMPARATORS: Record<string, string> = {
    eq: "=",
    ne: "!=",
    gt: ">",
    gte: ">=",
    lt: "<",
    lte: "<=",
  };

  /**
   * Translate a single field condition into Upstash filter syntax. Handles the
   * operator dicts, "in" arrays and "*" wildcard that mem0 supports — the old
   * version only emitted `key = value`, so anything else serialized to
   * `key = [object Object]` and Upstash rejected the query.
   */
  private buildFilterClause(key: string, value: any): string | undefined {
    // "*" means "the field must exist".
    if (value === "*") return `HAS FIELD ${key}`;

    // Array shorthand -> IN.
    if (Array.isArray(value)) {
      return `${key} IN (${value.map((v) => this.stringifyFilterValue(v)).join(", ")})`;
    }

    // Operator dict, e.g. { gte: 18, lte: 65 } — apply EVERY operator (AND).
    if (typeof value === "object") {
      const parts: string[] = [];
      for (const [op, opValue] of Object.entries(value)) {
        const comparator = UpstashVector.COMPARATORS[op];
        if (comparator) {
          parts.push(
            `${key} ${comparator} ${this.stringifyFilterValue(opValue)}`,
          );
        } else if (op === "in" && Array.isArray(opValue)) {
          parts.push(
            `${key} IN (${opValue.map((v) => this.stringifyFilterValue(v)).join(", ")})`,
          );
        } else if (op === "nin" && Array.isArray(opValue)) {
          parts.push(
            `${key} NOT IN (${opValue.map((v) => this.stringifyFilterValue(v)).join(", ")})`,
          );
        } else {
          throw new Error(this.unsupportedOperatorMessage(op, key));
        }
      }
      return parts.length > 0 ? parts.join(" AND ") : undefined;
    }

    // Scalar equality.
    return `${key} = ${this.stringifyFilterValue(value)}`;
  }

  private unsupportedOperatorMessage(op: string, key: string): string {
    return (
      `Unsupported Upstash filter operator '${op}' for field '${key}'. ` +
      `Supported operators: ${Object.keys(UpstashVector.COMPARATORS).join(", ")}, in, nin.`
    );
  }

  private matchesFilters(
    vector: Vector<UpstashMetadata>,
    filters?: SearchFilters,
  ): boolean {
    if (!filters) {
      return true;
    }

    return Object.entries(filters).every(([key, value]) => {
      if (value === undefined || value === null) {
        return true;
      }
      return this.matchesFieldCondition(vector.metadata ?? {}, key, value);
    });
  }

  /** Client-side equivalent of buildFilterClause, applying every operator. */
  private matchesFieldCondition(
    metadata: Record<string, any>,
    key: string,
    value: any,
  ): boolean {
    const fieldValue = metadata[key];

    if (value === "*") return key in metadata;
    if (Array.isArray(value)) return value.includes(fieldValue);

    if (value && typeof value === "object") {
      return Object.entries(value).every(([op, opValue]) => {
        switch (op) {
          case "eq":
            return fieldValue === opValue;
          case "ne":
            return fieldValue !== opValue;
          case "gt":
            return fieldValue > (opValue as any);
          case "gte":
            return fieldValue >= (opValue as any);
          case "lt":
            return fieldValue < (opValue as any);
          case "lte":
            return fieldValue <= (opValue as any);
          case "in":
            return Array.isArray(opValue) && opValue.includes(fieldValue);
          case "nin":
            return !Array.isArray(opValue) || !opValue.includes(fieldValue);
          default:
            throw new Error(this.unsupportedOperatorMessage(op, key));
        }
      });
    }

    return fieldValue === value;
  }
}
